SPARK-5017 [MLlib] - Use SVD to compute determinant and inverse of covariance matrix

tgaloppo opened the pull request:
MultivariateGaussian was calling both pinv() and det() on the covariance matrix, effectively performing two matrix decompositions.  Both values are now computed using the singular value decompositon. Both the pseudo-inverse and the pseudo-determinant are used to guard against singular matrices.

tgaloppo committed:
SPARK-5017 - Use SVD to compute determinant and inverse of covariance matrix.
Code was calling both det() and pinv(), effectively performing two matrix decompositions.
Futhermore, Breeze pinv() currently fails for singular matrices.

tgaloppo committed:
SPARK-5017 - Use SVD to compute determinant and inverse of covariance matrix

SparkQA commented:
[Test build #24987 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/24987/consoleFull) for   PR 3871 at commit [`fd9784c`](https://github.com/apache/spark/commit/fd9784cad41ee76b4711cb702dbbb7216965db89).
 * This patch merges cleanly.

SparkQA commented:
[Test build #24987 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/24987/consoleFull) for   PR 3871 at commit [`fd9784c`](https://github.com/apache/spark/commit/fd9784cad41ee76b4711cb702dbbb7216965db89).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/24987/
Test PASSed.

jkbradley commented:
@tgaloppo Could you please add a description?  It can be based off of the JIRA, just enough to cover the main points of the PR.  Thanks!

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,62 @@
+import org.apache.spark.mllib.util.MLUtils
+
+/*
+ * This class provides basic functionality for a Multivariate Gaussian (Normal) Distribution
```
Perhaps you could add a note here about how this behaves when sigma is singular, plus a reference like [http://en.wikipedia.org/wiki/Multivariate_normal_distribution#Degenerate_case]

The doc could be a short version of what you have below for ```calculateCovarianceConstants```

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,62 @@
-  private val U = math.pow(2.0 * math.Pi, -mu.length / 2.0) * math.pow(det(sigma), -0.5)
-    
+
+  private val (sigmaInv2, u) = calculateCovarianceConstants
```
Could you please add explicit types + documentation here for clarity?

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,62 @@
+    u * math.exp(delta.t * sigmaInv2 * delta)
+  }
+  
+  /*
```
Use ```/**``` for class/method documentation.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,62 @@
+   *  
+   * Both the determinant and the inverse can be computed from the singular value decomposition
+   * of sigma.  Noting that covariance matrices are always symmetric and positive semi-definite,
+   * we can use the eigendecomposition (breeze provides one specifically for symmetric matrices,
```
No need to comment on Breeze here; you can in the PR description if it's in question.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,62 @@
+   * To guard against singular covariance matrices, this method computes both the 
+   * pseudo-determinant and the pseudo-inverse (Moore-Penrose).  Singular values are considered
+   * to be non-zero only if they exceed a tolerance based on machine precision, matrix size, and
+   * relation to the maximum singular value (same tolerance used by, ie, Octave).
```
"ie" --> "e.g."

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,62 @@
+    val tol = MLUtils.EPSILON * max(d) * d.length
+    
+    // pseudo-determinant is product of all non-zero eigenvalues
+    val pdetSigma = (0 until d.length).map(i => if (d(i) > tol) d(i) else 1.0).reduce(_ * _)
```
More concise:
```
val pdetSigma = d.activeValuesIterator.filter(_ > tol).foldLeft(1.0)(_ * _)
```

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,62 @@
+    val pdetSigma = (0 until d.length).map(i => if (d(i) > tol) d(i) else 1.0).reduce(_ * _)
+    
+    // calculate pseudo-inverse by inverting all non-zero eigenvalues
+    val pinvS = new DBV((0 until d.length).map(i => if (d(i) > tol) (1.0 / d(i)) else 0.0).toArray)
```
This too can be more concise.
You generally do not need to use the ```(0 until length).map``` pattern unless you need the indices; it is easier to map the values of an array like d directly.

jkbradley commented:
@tgaloppo @mengxr  What are your thoughts about doing the computation in log space as much as possible, and then exponentiating at the end?  I'm mainly thinking about numerical stability, but I could imagine wanting to provide pdf() and logpdf() methods eventually.

jkbradley commented:
@tgaloppo The logic looks good; my comments are basically about clarity (except for the log space question).  Thanks for the PR!

jkbradley commented:
One more request: Could you please add a unit test with a singular matrix?  Thank you!  Perhaps in a new suite for MultivariateGaussian

tgaloppo committed:
Merge branch 'spark-5017' of https://github.com/tgaloppo/spark into spark-5017

SparkQA commented:
[Test build #24998 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/24998/consoleFull) for   PR 3871 at commit [`b4415ea`](https://github.com/apache/spark/commit/b4415ea70055e8ca2c0444cf964b696f0e1e410d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #24998 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/24998/consoleFull) for   PR 3871 at commit [`b4415ea`](https://github.com/apache/spark/commit/b4415ea70055e8ca2c0444cf964b696f0e1e410d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/24998/
Test PASSed.

tgaloppo commented:
@jkbradley I think performing the pdf calculation in log-space (and providing a logpdf() method) is a good idea.  Perhaps we can make this part of transitioning MultivariateGaussian to public scope?

tgaloppo committed:
SPARK-5017 - Fixed to use SVD to compute determinant and inverse of covariance matrix.  Previous code called both pinv() and det(), effectively performing two matrix decompositions.
Additionally, the pinv() implementation in Breeze is known to fail for singular matrices.

tgaloppo committed:
Added test suite for MultivariateGaussian, including test for degenerate case.

SparkQA commented:
[Test build #25001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25001/consoleFull) for   PR 3871 at commit [`d448137`](https://github.com/apache/spark/commit/d448137b739691c152dd981f136cef62b65d4e50).
 * This patch merges cleanly.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,69 @@
+    // This prevents any inverted value from exceeding (eps * n * max(d))^-1
+    val tol = MLUtils.EPSILON * max(d) * d.length
+    
+    // pseudo-determinant is product of all non-zero eigenvalues
```
"eigenvalues" --> "singular values"  (here and in the next comment on line 79)

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,69 @@
+    val tol = MLUtils.EPSILON * max(d) * d.length
+    
+    // pseudo-determinant is product of all non-zero eigenvalues
+    val pdetSigma = d.activeValuesIterator.filter(_ > tol).reduce(_ * _)
```
If all singular values are <= tol, then this will throw an UnsupportedOperationException.  Could you perhaps catch it and throw a more meaningful error if that happens?

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussianSuite.scala:
```diff
@@ -0,0 +1,61 @@
+  test("multivariate") {
+    val x = Vectors.dense(0.0, 0.0).toBreeze.toDenseVector
+    
+    val mu = Vectors.dense(0.0, 0.0).toBreeze. toDenseVector
```
typo: space between . and toDenseVector

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussianSuite.scala:
```diff
@@ -0,0 +1,61 @@
+  test("multivariate degenerate") {
+    val x = Vectors.dense(0.0, 0.0).toBreeze.toDenseVector
+    
+    val mu = Vectors.dense(0.0, 0.0).toBreeze. toDenseVector
```
typo: space between . and toDenseVector

SparkQA commented:
[Test build #25001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25001/consoleFull) for   PR 3871 at commit [`d448137`](https://github.com/apache/spark/commit/d448137b739691c152dd981f136cef62b65d4e50).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25001/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,69 @@
-   */
+import org.apache.spark.mllib.util.MLUtils
+
+/*
```
Use ```/**```

jkbradley commented:
@tgaloppo Thanks for the updates.  Sure, the log-space computation could be in another PR.

Just to make sure: Did you compute the PDF values in the tests using other software?

tgaloppo commented:
@jkbradley I used Octave's mvnpdf from the statistics package for the non-singular cases; it can not handle singular covariance matrices, so I was only able to recreate the function using octaves' pinv() function.

tgaloppo committed:
Catch potential exception calculating pseudo-determinant.
Style improvements.

SparkQA commented:
[Test build #25006 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25006/consoleFull) for   PR 3871 at commit [`dc3d0f7`](https://github.com/apache/spark/commit/dc3d0f7d70cfe54350a1ef70b20e9120b8937919).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25006 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25006/consoleFull) for   PR 3871 at commit [`dc3d0f7`](https://github.com/apache/spark/commit/dc3d0f7d70cfe54350a1ef70b20e9120b8937919).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25006/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussianSuite.scala:
```diff
@@ -0,0 +1,61 @@
+    val x = Vectors.dense(0.0, 0.0).toBreeze.toDenseVector
+    
+    val mu = Vectors.dense(0.0, 0.0).toBreeze.toDenseVector
+    var sigma = Matrices.dense(2, 2, Array(1.0, 0.0, 0.0, 1.0)).toBreeze.toDenseMatrix
```
Good catch with the other var; could you please fix this and the test above too?  var -> val

tgaloppo replied:
@jkbradley No problem; fixed.

jkbradley commented:
@tgaloppo thanks for verifying about the test pdf values

tgaloppo committed:
Moved some test values from var to val.

SparkQA commented:
[Test build #25016 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25016/consoleFull) for   PR 3871 at commit [`629d9d0`](https://github.com/apache/spark/commit/629d9d03e4bfd79410c8e1471e05986970022175).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25016 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25016/consoleFull) for   PR 3871 at commit [`629d9d0`](https://github.com/apache/spark/commit/629d9d03e4bfd79410c8e1471e05986970022175).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25016/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,74 @@
+ * This class provides basic functionality for a Multivariate Gaussian (Normal) Distribution. In
+ * the event that the covariance matrix is singular, the density will be computed in a
+ * reduced dimensional subspace under which the distribution is supported.
+ * (see http://en.wikipedia.org/wiki/Multivariate_normal_distribution#Degenerate_case)
```
For link, use double-brackets:
```
(See [[http://en.wikipedia.org/wiki/Multivariate_normal_distribution#Degenerate_case]])
```

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussianSuite.scala:
```diff
@@ -0,0 +1,61 @@
+
+class MultivariateGaussianSuite extends FunSuite with MLlibTestSparkContext {
+  test("univariate") {
+    val x = Vectors.dense(0.0).toBreeze.toDenseVector
```
It might be good to test at 2 points, rather than just 1.  (This tests the distribution constant but not inv(Sigma).)

jkbradley commented:
@tgaloppo 2 more small comments, but after those, I believe this will be ready.  Thanks!

tgaloppo committed:
Added additional points to tests in test suite.
Fixed comment in MultivariateGaussian

tgaloppo commented:
@jkbradley Good call on the test suite; I have added some non-center points to the tests.  I also added the brackets to the in-comment link.

SparkQA commented:
[Test build #25038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25038/consoleFull) for   PR 3871 at commit [`a5b8bc5`](https://github.com/apache/spark/commit/a5b8bc55e5f4abdc7c1e53bd3c32dac508e03f9f).
 * This patch merges cleanly.

jkbradley commented:
@tgaloppo Thanks for the updates!  LGTM after tests pass

CC: @mengxr

SparkQA commented:
[Test build #25038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25038/consoleFull) for   PR 3871 at commit [`a5b8bc5`](https://github.com/apache/spark/commit/a5b8bc55e5f4abdc7c1e53bd3c32dac508e03f9f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25038/
Test PASSed.

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussianSuite.scala:
```diff
@@ -0,0 +1,69 @@
+
+class MultivariateGaussianSuite extends FunSuite with MLlibTestSparkContext {
+  test("univariate") {
+    val x1 = Vectors.dense(0.0).toBreeze.toDenseVector
```
We can construct `BDV` and `BDM` directly.

tgaloppo committed:
MultivariateGaussian - minor optimization in density calculation

MultivariateGaussianSuite - Create Breeze vectors and matrices directly instead of through MLlib vectors/matrices.

tgaloppo commented:
@mengxr Changes made.

SparkQA commented:
[Test build #25069 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25069/consoleFull) for   PR 3871 at commit [`383b5b3`](https://github.com/apache/spark/commit/383b5b3c69189adf3a5d3dfd7e4857056bcd488c).
 * This patch merges cleanly.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -17,23 +17,84 @@
+
+  /**
+   * Compute distribution dependent constants:
+   *    rootSigmaInv = D^(-1/2) * U, where sigma = U * D * U.t
```
```D^(-1/2) * U``` should be ```U * D^(-1/2)```, right?  If so, it is an issue in both the comments & code below

tgaloppo replied:
U D^-1 * U.t => U D^(-1/2) D^(-1/2) U.t => (U D^(-1/2)) (D^(-1/2) U.t)
...both are U and D are symmetric, so...
(U D^(-1/2)) = (U.t (D^(-1/2)).t) => (D^(-1/2) U).t
and
(D^(-1/2) U.t) = (D^(-1/2) U)
thus
U D^-1 U.t => (D^(-1/2) U).t (D^(-1/2) U) 
... bringing in the delta we get
delta.t (D^(-1/2) U).t (D^(-1/2) U) delta
=> ((D^(-1/2) U) delta).t (D^(-1/2) U) delta = norm(D^(-1/2) U delta)^2
as indicated by @mengxr 

(phew, hope I did that OK! :) )

jkbradley replied:
In that case, U must not be symmetric.  But I must have been wrong about the orientation of U returned by the svd call.

tgaloppo replied:
I agree; I see no reason why U should always be symmetric (as you have demonstrated).  Before I roll back the change, I just want to make sure that I did not misinterpret @mengxr , and/or that we are not missing something that he sees.

jkbradley replied:
I think it's just a documentation error.  ```sigma = u * diag(d) * u.t``` should instead be ```sigma = u.t * diag(d) * u``` (for the value u being stored)

tgaloppo replied:
(u, d) is the eigendecomposition of sigma, so sigma = u * diag(d) * u^-1 ... but we have a special case since covariance matrices are always symmetric and positive semi-definite, in which case u * u.t = I, making it equivalent to the singular value decomposition... so sigma = u * diag(d) * u.t ... so in svd terms, v.t = u.t, then the inverse is v * inv(diag(d)) * u.t = u * inv(diag(d)) * u.t ...

Have I lost my bearings?

jkbradley replied:
Ugh, no, I have.  I was confused about which were inverses, and what you wrote looks perfectly fine.  Sorry for the trouble!  I'll remove the comments.d

SparkQA commented:
[Test build #25069 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25069/consoleFull) for   PR 3871 at commit [`383b5b3`](https://github.com/apache/spark/commit/383b5b3c69189adf3a5d3dfd7e4857056bcd488c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25069/
Test PASSed.

jkbradley commented:
@tgaloppo Now that my confusion is over...LGTM  Thanks very much!
CC: @mengxr

mengxr commented:
Merged into master. Thanks!
